File: apps/grokking-ml-python/src/grokking_ml/ch05_perceptron.py
"""第 5 章: パーセプトロン。

点を直線で 2 クラスに分ける。予測が外れた点だけを使って境界線を動かす
パーセプトロントリックを実装する。
"""

from __future__ import annotations

import random
from collections.abc import Sequence
from dataclasses import dataclass

Point = Sequence[float]


@dataclass(frozen=True)
class Perceptron:
    """線形分類器 score = w・x + bias。score >= 0 なら 1、そうでなければ 0。"""

    weights: tuple[float, ...]
    bias: float

    def score(self, point: Point) -> float:
        return self.bias + sum(w * x for w, x in zip(self.weights, point))

    def predict(self, point: Point) -> int:
        return 1 if self.score(point) >= 0 else 0


def perceptron_trick(
    model: Perceptron,
    point: Point,
    label: int,
    learning_rate: float = 0.01,
) -> Perceptron:
    """パーセプトロントリック。誤分類した点だけモデルを動かす。"""
    error = label - model.predict(point)
    if error == 0:
        return model
    weights = tuple(w + learning_rate * error * x for w, x in zip(model.weights, point))
    return Perceptron(weights, model.bias + learning_rate * error)


def perceptron_error(model: Perceptron, point: Point, label: int) -> float:
    """1 点分の誤差。正しく分類していれば 0、誤っていればスコアの絶対値。"""
    if model.predict(point) == label:
        return 0.0
    return abs(model.score(point))


def mean_perceptron_error(
    model: Perceptron,
    points: Sequence[Point],
    labels: Sequence[int],
) -> float:
    """全点の平均誤差。"""
    total = sum(perceptron_error(model, point, label) for point, label in zip(points, labels))
    return total / len(points)


def accuracy(model: Perceptron, points: Sequence[Point], labels: Sequence[int]) -> float:
    """正解率。"""
    correct = sum(1 for point, label in zip(points, labels) if model.predict(point) == label)
    return correct / len(points)


def perceptron_algorithm(
    points: Sequence[Point],
    labels: Sequence[int],
    learning_rate: float = 0.01,
    epochs: int = 1000,
    seed: int = 0,
) -> tuple[Perceptron, list[float]]:
    """パーセプトロンアルゴリズム。モデルとエポックごとの平均誤差を返す。"""
    rng = random.Random(seed)
    dimensions = len(points[0])
    model = Perceptron(tuple(0.0 for _ in range(dimensions)), 0.0)
    errors: list[float] = []
    for _ in range(epochs):
        errors.append(mean_perceptron_error(model, points, labels))
        i = rng.randrange(len(points))
        model = perceptron_trick(model, points[i], labels[i], learning_rate)
    return model, errors

File: apps/grokking-ml-fsharp/src/GrokkingMl/Ch05Perceptron.fs
/// 第 5 章: パーセプトロン。
/// 点を直線で 2 クラスに分ける。予測が外れた点だけを使って境界線を動かす
/// パーセプトロントリックを実装する。
module GrokkingMl.Ch05Perceptron

open System

/// 特徴量ベクトル。
type Point = float list

/// 線形分類器 score = w・x + bias。score >= 0 なら 1、そうでなければ 0。
type Perceptron =
    { Weights: float list
      Bias: float }

/// 重み付き和。
let score (model: Perceptron) (point: Point) =
    List.map2 (*) model.Weights point
    |> List.sum
    |> (+) model.Bias

/// スコアの符号による分類。境界線上は正のクラスとして扱う。
let predict (model: Perceptron) (point: Point) = if score model point >= 0.0 then 1 else 0

/// パーセプトロントリック。誤分類した点だけモデルを動かす。
let perceptronTrick (learningRate: float) (model: Perceptron) (point: Point) (label: int) =
    let error = float (label - predict model point)

    if error = 0.0 then
        model
    else
        { Weights = List.map2 (fun w x -> w + learningRate * error * x) model.Weights point
          Bias = model.Bias + learningRate * error }

/// 1 点分の誤差。正しく分類していれば 0、誤っていればスコアの絶対値。
let perceptronError (model: Perceptron) (point: Point) (label: int) =
    if predict model point = label then
        0.0
    else
        abs (score model point)

/// 全点の平均誤差。
let meanPerceptronError (model: Perceptron) (points: Point list) (labels: int list) =
    List.map2 (perceptronError model) points labels
    |> List.average

/// 正解率。
let accuracy (model: Perceptron) (points: Point list) (labels: int list) =
    List.map2 (fun point label -> if predict model point = label then 1.0 else 0.0) points labels
    |> List.average

/// パーセプトロンアルゴリズム。モデルとエポックごとの平均誤差を返す。
let perceptronAlgorithm
    (learningRate: float)
    (epochs: int)
    (seed: int)
    (points: Point list)
    (labels: int list)
    =
    let rng = Random(seed)
    let dimensions = List.length (List.head points)
    let initial = { Weights = List.replicate dimensions 0.0; Bias = 0.0 }

    let step (model, errors) _ =
        let errors = meanPerceptronError model points labels :: errors
        let i = rng.Next(List.length points)
        let model = perceptronTrick learningRate model points[i] labels[i]
        (model, errors)

    let model, errors = List.fold step (initial, []) [ 1..epochs ]
    (model, List.rev errors)
